File: full-stack-asset-transfer-guide/applications/trader-typescript/src/contract.ts
/*
 * Copyright IBM Corp. All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { CommitError, Contract, StatusCode } from '@hyperledger/fabric-gateway';
import { TextDecoder } from 'util';

const RETRIES = 2;

const utf8Decoder = new TextDecoder();

export interface Asset {
    ID: string;
    Color: string;
    Size: number;
    Owner: string;
    AppraisedValue: number;
}

export type AssetCreate = Omit<Asset, 'Owner'> & Partial<Asset>;
export type AssetUpdate = Pick<Asset, 'ID'> & Partial<Omit<Asset, 'Owner'>>;

/**
 * AssetTransfer presents the smart contract in a form appropriate to the business application. Internally it uses the
 * Fabric Gateway client API to invoke transaction functions, and deals with the translation between the business
 * application and API representation of parameters and return values.
 */
export class AssetTransfer {
    readonly #contract: Contract;

    constructor(contract: Contract) {
        this.#contract = contract;
    }

    async createAsset(asset: AssetCreate): Promise<void> {
        await this.#contract.submit('CreateAsset', {
            arguments: [JSON.stringify(asset)],
        });
    }

    async getAllAssets(): Promise<Asset[]> {
        const result = await this.#contract.evaluate('GetAllAssets');
        if (result.length === 0) {
            return [];
        }

        return JSON.parse(utf8Decoder.decode(result)) as Asset[];
    }

    async readAsset(id: string): Promise<Asset> {
        const result = await this.#contract.evaluate('ReadAsset', {
            arguments: [id],
        });
        return JSON.parse(utf8Decoder.decode(result)) as Asset;
    }

    async updateAsset(asset: AssetUpdate): Promise<void> {
        await submitWithRetry(() => this.#contract.submit('UpdateAsset', {
            arguments: [JSON.stringify(asset)],
        }));
    }

    async deleteAsset(id: string): Promise<void> {
        await submitWithRetry(() => this.#contract.submit('DeleteAsset', {
            arguments: [id],
        }));
    }

    async assetExists(id: string): Promise<boolean> {
        const result = await this.#contract.evaluate('AssetExists', {
            arguments: [id],
        });
        return utf8Decoder.decode(result).toLowerCase() === 'true';
    }

    async transferAsset(id: string, newOwner: string, newOwnerOrg: string): Promise<void> {
        console.log(`transferring asset '${id}' to ${newOwner}, ${newOwnerOrg}`);
        // TODO: implement the transferAsset() function.
        // TODO: submit a 'TransferAsset' transaction, requiring [id, newOwner, newOwnerOrg] arguments.
        return Promise.reject(new Error('TODO: implement the contract.ts transferAsset() function.'));
    }
}

async function submitWithRetry<T>(submit: () => Promise<T>): Promise<T> {
    let lastError: unknown | undefined;

    for (let retryCount = 0; retryCount < RETRIES; retryCount++) {
        try {
            return await submit();
        } catch (err: unknown) {
            lastError = err;
            if (err instanceof CommitError) {
                // Transaction failed validation and did not update the ledger. Handle specific transaction validation codes.
                if (err.code === StatusCode.MVCC_READ_CONFLICT) {
                    continue; // Retry
                }
            }
            break; // Failure -- don't retry
        }
    }

    throw lastError;
}

File: off_chain_data/application-go/contract/contract.go
package contract

import (
	"encoding/json"
	"strconv"

	"github.com/hyperledger/fabric-gateway/pkg/client"
)

type AssetTransferBasic struct {
	contract *client.Contract
}

func NewAssetTransferBasic(contract *client.Contract) *AssetTransferBasic {
	return &AssetTransferBasic{contract}
}

func (atb *AssetTransferBasic) CreateAsset(anAsset Asset) error {
	if _, err := atb.contract.Submit(
		"CreateAsset",
		client.WithArguments(
			anAsset.ID,
			anAsset.Color,
			strconv.FormatUint(anAsset.Size, 10),
			anAsset.Owner,
			strconv.FormatUint(anAsset.AppraisedValue, 10),
		)); err != nil {
		return err
	}
	return nil
}

func (atb *AssetTransferBasic) TransferAsset(id, newOwner string) (string, error) {
	result, err := atb.contract.Submit(
		"TransferAsset",
		client.WithArguments(
			id,
			newOwner,
		),
	)
	if err != nil {
		return "", err
	}

	return string(result), nil
}

func (atb *AssetTransferBasic) DeleteAsset(id string) error {
	if _, err := atb.contract.Submit(
		"DeleteAsset",
		client.WithArguments(
			id,
		),
	); err != nil {
		return err
	}
	return nil
}

func (atb *AssetTransferBasic) GetAllAssets() ([]Asset, error) {
	assetsRaw, err := atb.contract.Evaluate("GetAllAssets")
	if err != nil {
		return nil, err
	}

	if len(assetsRaw) == 0 {
		return nil, nil
	}

	assets := []Asset{}
	if err := json.Unmarshal(assetsRaw, &assets); err != nil {
		return nil, err
	}

	return assets, nil
}
